### DISCUSSION TITLE
Make an `HttpWebRequest` only during normal builds and not design-time ones

### DISCUSSION THREAD

**Discussion by LaPeste** [dotnet/roslyn#65506] [Q&A] [answered]
Similarly to a [previous question](https://github.com/dotnet/roslyn/discussions/50568) that was asked here, I'd like the SG to make an `HttpWebRequest`.
In the linked conversation it is strongly advised to avoid this kind of IO operation because the SG should only be given a sub-ms budget in order to leave the IDE responsive.
It'd be fully acceptable to do this IO operation only during normal builds and not design-time builds.

Do you have any advice on how to achieve this?

**Comment by CyrusNajmabadi** [comment, marked as answer]
This isn't supported in any scenario. The problem here is that this makes builds non deterministic which violates a core design principle we have. I.e. we might not run the generator again if we believe nothing would change. But we have no idea that results would be different because you're hitting a web server. 

The design to use is to pull down the data locally first and add as an 'additional file' which the generator can then see.

**Reply by jaredpar** [reply]
100% correct. If we could ban I/O in generators we would. If we get the capability in the future to do so we will. 

All I/O should be done in preceeding build tasks and passed to compilers via additional files.

**Reply by LaPeste** [reply]
Thank you for the input. We're looking into the possibility of splitting our logic and have an msbuild task doing the IO work.
